README.md
# LoveConsole

A simple to use and powerful command Console to your love games.

---

## Features

- Draggable window
- Custom command support (supporting commands with parameters, without changing the source)
- Simple color support
- Easy to install and setup.
- No need external files to work.
- Window resizing and auto breaking text

### Planned:

- [ ] More default commands

---

## How to install

> ### **[1] -** Copy the console library to your project folder

> ### **[2] -** require it with `console = require 'console'`

> ### **[3] -** Put every function on specific callback. Example : console:update(dt) to to love.update(dt) callback.

---

## Examples of use

### **[Example 1] |** using with [hump gamestate](https://github.com/vrld/hump/tree/master)

Instead of calling every function of the console library in every state you create, this example shows you how to create a global console.

```lua
function love.load()
    -- require the console library --
    console = require 'console'
    gamestate = require 'gamestate'

    -- create a new console at x: 90, y: 90
    console:init()

    -- states list (cool for organization :D) --
    states = {
        myState = require 'states.mystate'
    }
    
    -- register the necessary events --
    gamestate.registerEvents({'update', 'textinput', 'keypressed', 'mousepressed', 'mousereleased'})
    -- change the state --
    gamestate.switch(states.myState)
end

function love.draw()
    -- draw all the state stuff first --
    gamestate.current():draw()
    -- after drawing the state stuff, draw the console itself --
    console:render()
end

function love.update(elapsed)
    console:update()
end

function love.textinput(text)
    console:textinput(text)
end

function love.keypressed(k)
    console:keypressed(k)
end

function love.mousepressed(x, y, button)
    console:mousepressed(x, y, button)
end

function love.mousereleased(x, y, button)
    console:mousereleased(x, y, button)
end
```

> **OBS:** This method don't work for `love.graphics.present()` function. Everything draw on the present() method will appear on the top

> **OBS 2:** This example shows you how to setup the console globally, removing the need to install it on every state you create and want use.

---

### **[Example 2] |** creating custom commands (Part 1 - The basic command)

This example will show you how to create a basic command *without any arguments*. It uses a function called `console:registerCommand:(<commandName>:string, <helpDescription>:string, <commandFunction>:function)`

> **OBS :** This example also uses another command, the `console:trace()` command, it will be explained on the functions section.

```lua
function love.load()
    -- require the console library --
    console = require 'console'
    -- create a new console at x: 90, y: 90
    console:init()

    console:registerCommand("myCoolCommandName", "This is the help message, it will appear with help command" function()
        console:trace("This is my cool command")
    end)
end
```

and this is the result : 

![Alt text](assets/Screenshot_1.png)

and when we run the command on the console:

![Alt text](assets/Screenshot_2.png)

### **[Example 3] |** creating custom commands (Part 2 - Commands with arguments)

Is the same process to create commands but this time, we include the arguments we want as parameters on the function

```lua
function love.load()
    -- require the console library --
    console = require 'console'
    -- create a new console at x: 90, y: 90
    console:init()

    console:registerCommand("myCoolCommandWithArgument", "This is the help message, it will appear with help command", 0, function(argument)
        console:trace("This is my cool command")
        console:trace("And this is my argument" .. argument)
    end)
end
```

and this is the result when we execute it on the console :

![Alt text](assets/Screenshot_3.png)

## Function list

---

### **console:registerCommand()**

---

Used to register a custom command to commands database.

Usage : `console:registerCommands(<commandName>, <helpDescription>, <priorityLevel>, <functionExec>)`

| Parameters  | Type | optional | Description |
| ------------- | ------------- | ------------- | ------------- |
| commandName | `string` | no | The name of your command |
| helpDescription | `string` | yes  | The description of your command |
| priorityLevel | `number` | yes | Define the priority color on the `help` command|
| functionExec | `function` | no | The lua function you want run when the command is executed|

---

### **console:rebind()**

---

Used to rebind some keyboard keys functions

Usage : `console:rebind(<keys>)`

| Parameters  | Type | optional | Description |
| ------------- | ------------- | ------------- | ------------- |
| keys | `table` | no | The keys table you want rebind |

here the list of the possible values to use in this table:

| Keys | value |
| ---- | ----- |
| submit | return |
| open | f1 |
| removeChar | backspace |
| previousCommand | up |
| nextcommand | down |

example :

```lua
console:rebind({
    submit = "return",
    open = "f1",
    removeChar = "backspace",
    previousCommand = "up"
    nextcommand = "down"
})
```

> **OBS :** Please insert a valid key constant, if you want check the official [LOVE keys Constant reference](https://love2d.org/wiki/KeyConstant)

---

### **console:trace()**

---

Used to write a message on the console display.

Usage: `console:trace(<message>, <level>)`

| Parameters  | Type | optional | Description |
| ------------- | ------------- | ------------- | ------------- |
| message | `string` | no | The message you want write |

---

### **console:setTheme()**

---

Used to change the colors of the console including  the text colors

Usage : `console:setTheme(_theme)`

| Parameters  | Type | optional | Description |
| ------------- | ------------- | ------------- | ------------- |
| theme | `table` | no | The theme data table |

here the list of the possible values to use in this table:

| Key | value |
| ---- | ----- |
| bg | table of colors (0 - 255) |
| fg | table of colors (0 - 255) |
| textColor | table of colors (0 - 255) |


**Example** :
```lua
function love.load()
    loveconsole:setTheme({
        bg = {255, 255, 255},
        fg = {128, 128, 128},
        textColor = {
            {255, 255, 255},
            {128, 128, 128},
            {0, 0, 0}
        }
    })
end
```

CHANGELOG.txt
[1.0]

- first release

[1.1]

- Legacy code moved to folder `Legacy`
- LoveConsole rewriteen from scratch with better functions

loveconsole.lua
console = {}

local lines = 0

function console:init()
    utf8 = require 'utf8'

    love.keyboard.setKeyRepeat(true)
    self.x = 90
    self.y = 90
    self.w = 640
    self.h = 480
    self.binds = {}
    self.binds.submit = "return"
    self.binds.removeChar = "backspace"
    self.binds.open = "f1"
    self.meta = {}
    self.terminal = {}
    self.commands = {
        {
            name = "help",
            description = "this is the help message",
            run = function()
                console:trace("Showing " .. #self.commands .. " commands", 0)
                console:trace("===============================================", 0)
                for h = 1, #self.commands, 1 do
                    console:trace(self.commands[h].name .. " | " .. self.commands[h].description)
                end
            end
        },
        {
            name = "trace",
            description = "this is the help message",
            run = function(...)
                local data = {...}
                console:trace("[" .. os.date("%H:%M:%S") .. "] "..  table.concat(data, " "), 2)
            end
        },
        {
            name = "color",
            description = "change the color",
            run = function(color)
                if color == nil then
                    console:trace("[:ERROR:] This command require argument")
                    return
                end
                if tonumber(color) <= 1 then
                    self.theme.textColorID = 1
                elseif tonumber(color) >= #self.theme.textColors then
                    self.theme.textColorID = #self.theme.textColors
                else
                    self.theme.textColorID = tonumber(color)
                end
            end
        }
    }
    self.meta.objects = {}
    self.meta.form = {}
    self.meta.form.command = ""
    self.meta.isEnable = false
    self.meta.dragging = {}
    self.meta.dragging.active = false
    self.meta.dragging.diffX = 0
    self.meta.dragging.diffY = 0
    self.theme = {}
    self.theme.bg = {116, 155, 76}
    self.theme.fg = {67,90, 43}
    self.theme.fontsize = 12
    self.theme.fontpath = nil
    self.theme.textColors = {
        {255, 255, 255},
        {128, 128, 128},
        {0, 0, 0},
        {255, 0, 0},
        {128, 0, 0},
        {0, 255, 0},
        {0, 128, 0},
        {0, 0, 255},
        {0, 0, 128},
        {255, 255, 0},
        {128, 128, 0},
        {0, 255, 255},
        {0, 128, 128},
        {255, 0, 255},
        {128, 0, 128}
    }
    self.theme.textColorID = 1
    if self.theme.fontpath == nil then
        self.meta.objects.font = love.graphics.newFont(self.theme.fontsize)
    else
        self.meta.objects.font = love.graphics.newFont(self.theme.fontpath, self.theme.fontsize)
    end
    self.registry = {}
    self.registry.showTraceback = false
    love.graphics.setFont(self.meta.objects.font)
end

function console:render()
    if not self then
        error("[:ERROR:] Console is not initialized")
    end
    if self.meta.isEnable then
        lines = math.ceil(self.meta.objects.font:getWidth(self.meta.form.command) / (self.w - 20))
        --% The back of the window %--
        love.graphics.setColor(self.theme.bg[1] / 255, self.theme.bg[2] / 255, self.theme.bg[3] / 255)
        if lines > 1 then
            love.graphics.rectangle("fill", self.x, self.y, self.w, self.h + (lines * self.meta.objects.font:getHeight()) - 14, 5, 5)
        else
            love.graphics.rectangle("fill", self.x, self.y, self.w, self.h , 5, 5)
        end
        love.graphics.setColor(1, 1, 1, 1)
        --% The hole of the window %--  
        love.graphics.setColor(self.theme.fg[1] / 255, self.theme.fg[2] / 255, self.theme.fg[3] / 255)
        love.graphics.rectangle("fill", self.x + 10, self.y + 10, self.w - 20, self.h - 60, 5, 5)
        love.graphics.setColor(1, 1, 1, 1)
        --% TheInput hole %--  
        love.graphics.setColor(self.theme.fg[1] / 255, self.theme.fg[2] / 255, self.theme.fg[3] / 255)
        --print(self.meta.objects.font:getHeight(self.meta.form.fullCommand))
        if lines > 1 then
            love.graphics.rectangle("fill", self.x + 10, (self.y + self.h) - 40, self.w - 20, (lines * self.meta.objects.font:getHeight()) + 14 , 5, 5)
        else
            love.graphics.rectangle("fill", self.x + 10, (self.y + self.h) - 40, self.w - 20, 32 , 5, 5)
        end
        love.graphics.setColor(1, 1, 1, 1)
        --% render the text %--
        love.graphics.printf(self.meta.form.command .. "_", self.x + 13, (self.y + self.h) - 32, self.w - 20, "left")
        --% render the terminal stuff %--
        --print(self.theme.textColors[self.theme.textColorID][1], self.theme.textColors[self.theme.textColorID][2], self.theme.textColors[self.theme.textColorID][3])
        love.graphics.setColor(self.theme.textColors[self.theme.textColorID][1] / 255, self.theme.textColors[self.theme.textColorID][2] / 255, self.theme.textColors[self.theme.textColorID][3] / 255)
        love.graphics.printf(table.concat(self.terminal, "\n"), self.x + 10, self.y + 10, self.w - 20, "left")
        love.graphics.setColor(1, 1, 1, 1)
    end
end

function console:update(elapsed)
    if not self then
        error("[:ERROR:] Console is not initialized")
    end
    local width, terminalLines = self.meta.objects.font:getWrap(table.concat(self.terminal, "\n"), self.w - 20)
    if #terminalLines * self.meta.objects.font:getHeight() > self.h - 60  then
        table.remove(self.terminal, 1)
    end
    if self.meta.isEnable then
        if self.meta.dragging.active then
            self.x = love.mouse.getX() - self.meta.dragging.diffX
            self.y = love.mouse.getY() - self.meta.dragging.diffY
        end
    end
end

function console:textinput(t)
    if not self then
        error("[:ERROR:] Console is not initialized")
    end
    if self.meta.isEnable then
        if t ~= self.binds.open then
            self.meta.form.command = self.meta.form.command .. t
        end
    end
end

function console:keypressed(k)
    if not self then
        error("[:ERROR:] Console is not initialized")
    end
    if k == self.binds.open then
        if self.meta.isEnable then
            self.meta.isEnable = false
        else
            self.meta.isEnable = true
        end
    end
    if k == self.binds.removeChar then
        local byteoffset = utf8.offset(self.meta.form.command, -1)
        if byteoffset then
            self.meta.form.command = string.sub(self.meta.form.command, 1, byteoffset - 1)
        end
    end
    if k == self.binds.submit then
        console:submit()
    end
end

function console:mousepressed(x, y, button)
    if not self then
        error("[:ERROR:] Console is not initialized")
    end
    if button == 1
    and x > self.x and x < self.x + self.w
    and y > self.y and y < self.y + self.h
    and self.meta.isEnable
    then
        self.meta.dragging.active = true
        self.meta.dragging.diffX = x - self.x
        self.meta.dragging.diffY = y - self.y
    else
        self.meta.dragging.active = false
    end
end

function console:mousereleased(x, y, button)
    if not self then
        error("[:ERROR:] Console is not initialized")
    end
    if button == 1 then
        self.meta.dragging.active = false
    end
end

--------------------=[ Special Functions ]=--------------------------


--- run a command on the console
---@param cmd string
function console:run(cmd)
    local tkn = _tokenize(cmd, " ")
    for _, commands in ipairs(self.commands) do
        if commands.name == tkn[1] then
            table.remove(tkn, 1)
            sucess, error = pcall(commands.run, unpack(tkn))
            if self.registry.showTraceback then
                console:trace(error)
            end
        end
    end
    tkn = {}
end

function console:submit()
    if self.meta.isEnable then
        if self.meta.form.command ~= "" then
            table.insert(self.terminal, self.meta.form.command)
            console:run(self.meta.form.command)
            self.meta.form.command = ""
            --print(debug.getTableContent(self.terminal))
        end
    end
end

--- Create your custom commands
---@param commandName string
---@param helpDescription string
---@param priorityLevel number
---@param func function
function console:registerCommand(commandName, helpDescription, func)
    Command = {}
    Command.name = commandName or "defaultCommand"
    Command.description = helpDescription or ""
    Command.run = func or function()
        console:trace("please replace it with console:registerCommand()", 2)
    end
    table.insert(self.commands, Command)
end

--- rebind keys
---@param keys table
function console:rebind(keys)
    self.binds.submit = keys.submit or "return"
    self.binds.open = keys.open or "f1"
    self.binds.removeChar = keys.removeChar or "backspace"
end

--- enable the console directly from code
---@param bool boolean
function console:enable(bool)
    self.meta.isEnable = bool
end

--- echo to the console
---@param text string
function console:trace(text)
    table.insert(self.terminal, text)
end

--- Edt the console theme
---@param theme table
function console:setTheme(theme)
    self.theme.bg = theme.bg or self.theme.bg
    self.theme.fg = theme.fg or self.theme.fg
    self.theme.textColors = theme.textColor or self.theme.textColors
end

----------------------=[ Local functions ]=------------------------

function _tokenize(inputstr, sep)
    if sep == nil then
        sep = "%s"
    end
    local t = {}
    for str in string.gmatch(inputstr, "([^" .. sep .. "]+)") do
        table.insert(t, str)
    end
    return t
end

return console

Legacy/console.lua
console = {
    author = "StrawberryChocolate",
    Version = "1.0"
}

utf8 = require 'utf8'

---  Create a new console (if nil it will create a console at x: 90 y: 90)
---@param x number
---@param y number
function console:new(x, y)
    self.x = x or 90
    self.y = y or 90
    self.w = 640
    self.h = 480
    self.binds = {}
    self.binds.submit = "return"
    self.binds.open = "f1"
    self.binds.removeChar = "backspace"
    self.binds.previousCommand = "up"
    self.binds.nextCommand = "down"
    self.meta = {}
    self.meta.terminal = {}
    self.meta.commands = {
        -- pre-registred commands --
        {
            name = "clear",
            description = "clear the screen",
            priority = 0,
            func = function()
                self.meta.terminal = {}
            end
        },
        {
            name = "console",
            description = "this is the help message",
            priority = 2,
            func = function(...)
                local tokens = cunpack(cpack({...}))
                if tokens[1] == "mode" then
                    if tokens[2] ~= nil then
                        if type(tonumber(tokens[2])) == "number" then
                            if tokens[3] ~= nil then
                                if type(tonumber(tokens[3])) == "number" then
                                    self.w = tonumber(tokens[2])
                                    self.h = tonumber(tokens[3])
                                    console:trace("Changed size!", 3)
                                else
                                    console:trace("this argument require a 'number'", 2)
                                end
                            else
                                console:trace("invalid command argument", 2)
                            end
                        else
                            console:trace("this argument require a 'number'", 2)
                        end
                    else
                        console:trace("invalid command argument", 2)
                    end
                elseif tokens[1] == "position" then
                    if tokens[2] ~= nil then
                        if type(tonumber(tokens[2])) == "number" then
                            if tokens[3] ~= nil then
                                if type(tonumber(tokens[3])) == "number" then
                                    self.x = tonumber(tokens[2])
                                    self.y = tonumber(tokens[3])
                                    console:trace("Changed size!", 3)
                                else
                                    console:trace("this argument require a 'number'", 2)
                                end
                            else
                                console:trace("invalid command argument", 2)
                            end
                        else
                            console:trace("this argument require a 'number'", 2)
                        end
                    else
                        console:trace("invalid command argument", 2)
                    end
                else
                    console:trace("invalid command argument", 2)
                end
            end
        },
        {
            name = "engine",
            description = "this is the help message",
            priority = 2,
            func = function(...)
                local tokens = cunpack(cpack({...}))
                if tokens == nil then
                    console:trace("This command require an argument to run correctly", 2)
                else
                    if tokens[1] == "draw" then
                        if tokens[2] == "wireframe" then
                            if tokens[3] == nil then
                                console:trace("This command require an argument to run correctly :: required boolean", 2)
                            else
                                love.graphics.setWireframe(toboolean(tokens[3]))
                            end
                        else
                            console:trace("Invalid property", 2)
                        end
                    else
                        console:trace("Invalid property", 2)
                    end
                end
            end
        },
        {
            name = "help",
            description = "this is the help message",
            priority = 0,
            func = function()
                console:trace("Showing " .. #self.meta.commands .. " commands", 0)
                console:trace("----------------------------------------------", 0)
                for h = 1, #self.meta.commands, 1 do
                    console:trace(self.meta.commands[h].name .. " | " .. self.meta.commands[h].description, self.meta.commands[h].priority)
                end
            end
        },
        {
            name = "echo",
            description = "show a message on the console",
            priority = 0,
            func = function(...)
                local tkns = cpack({...})
                if ... == nil then
                    console:trace("no output", 2)
                else
                    console:trace(table.concat(cunpack(tkns), " "), 3)
                end
            end
        },
        {
            name = "event",
            description = "execute a event",
            priority = 2,
            func = function(arg)
                if arg == nil then
                    console:trace("Failed to run 'event' command :: no arguments found :: usage | app <argument>" , 2)
                else
                    if arg == "quit" then
                        love.event.quit()
                    elseif arg == "restart" or arg == "reset" then
                        love.event.quit("restart")
                    else
                        console:trace("Invalid argument : " .. "'" .. arg .. "'", 2)
                    end
                end
            end
        },
        {
            name = "window",
            description = "control window",
            priority = 2,
            func = function(...)
                local tokens = cunpack(cpack({...}))
                if ... == nil then
                    console:trace("This command require arguments", 2)
                else
                    if tokens[1] == "title" then
                        if tokens[2] == nil then
                            console:trace("This command require arguments", 2)
                        else
                            love.window.setTitle(tostring(tokens[2]))
                        end
                    elseif tokens[1] == "position" then
                        if type(tonumber(tokens[2])) == "number" then
                            if type(tonumber(tokens[3])) == "number" then
                                love.window.setPosition(tokens[2], tokens[3])
                            else
                                console:trace("This command require arguments in number type", 2)
                            end
                        else
                            console:trace("This command require arguments in number type", 2)
                        end
                    else
                        console:trace("This command require arguments", 2)
                    end
                end
            end
        },
    }
    self.meta.consoleEnable = false
    self.meta.dragging = {}
    self.meta.dragging.active = false
    self.meta.dragging.diffX = 0
    self.meta.dragging.diffY = 0
    self.meta.currentItem = 1
    self.meta.maxRender = 26
    self.meta.fontSize = 12
    self.meta.textCommand = ""
    self.meta.font = love.graphics.newFont(self.meta.fontSize)
    self.meta.commandHistory = {}
    self.meta.commandHistorySelection = 1
    console:announce("LoveConsole v" .. console.Version)
    console:announce("By " .. console.author)
    console:announce("Love version : " .. string.format("%d.%d.%d - %s", love.getVersion()))
    console:trace("[WARNING] You are using a deprecated version of LoveConsole.", 1)
    console:trace("[WARNING] Please use the lastest version.", 1)
    console:trace("Loaded " .. #self.meta.commands .. " commands", 3)
end


function console:render()
    local textY = self.y + 20
    if self.meta.consoleEnable then
        love.graphics.setColor(100 / 255, 100 / 255, 100 / 255)
        love.graphics.rectangle("fill", self.x, self.y, self.w, self.h, 10, 10)
        love.graphics.setColor(255 / 255, 255 / 255, 255 / 255)
        love.graphics.setColor(40 / 255, 40 / 255, 40 / 255)
        love.graphics.rectangle("fill", self.x + 10, self.y + 20, self.w - 20, self.h - 30, 10, 10)
        love.graphics.setColor(255 / 255, 255 / 255, 255 / 255, 255 / 255)
        love.graphics.print("Console", self.x + 15, self.y + 5)
        love.graphics.print("> " .. tostring(self.meta.textCommand) .. "_", self.meta.font, self.x + 13, self.y + (self.h - 32))
        for ti = 1, #self.meta.terminal, 1 do
            if #self.meta.terminal > 0 then
                if self.meta.terminal[ti].type == "trace" then
                    love.graphics.setColor(self.meta.terminal[ti].levelColor)
                    love.graphics.print("] " .. self.meta.terminal[ti].text, self.meta.font, self.x + 13, textY)
                    love.graphics.setColor(255 / 255, 255 / 255, 255 / 255, 255 / 255)
                elseif self.meta.terminal[ti].type == "announce" then
                    love.graphics.setColor(77 / 255, 33 / 255, 8 / 255)
                    love.graphics.print("] " .. self.meta.terminal[ti].text, self.meta.font, self.x + 13, textY + 1)
                    love.graphics.setColor(255 / 255, 255 / 255, 255 / 255, 255 / 255)
                    love.graphics.setColor(203 / 255, 138 / 255, 76 / 255)
                    love.graphics.print("] " .. self.meta.terminal[ti].text, self.meta.font, self.x + 13, textY - 1)
                    love.graphics.setColor(255 / 255, 255 / 255, 255 / 255, 255 / 255)
                    love.graphics.setColor(155 / 255, 102 / 255, 51 / 255)
                    love.graphics.print("] " .. self.meta.terminal[ti].text, self.meta.font, self.x + 13, textY)
                    love.graphics.setColor(255 / 255, 255 / 255, 255 / 255, 255 / 255)
                else
                    love.graphics.setColor(255 / 255, 255 / 255, 255 / 255, 255 / 255)
                    love.graphics.print("] " .. self.meta.terminal[ti], self.meta.font, self.x + 13, textY)
                end
                textY = textY + self.meta.fontSize
            end
        end
        love.graphics.setColor(255 / 255, 255 / 255, 255 / 255)
    end
end


function console:update(elapsed)
    --[[
    if self.meta.currentItem < 1 then
        self.meta.currentItem = 1
    end
    if self.meta.currentItem > #self.meta.terminal then
        self.meta.currentItem = #self.meta.terminal
    end
    ]]--
    if #self.meta.terminal > 28 then
        table.remove(self.meta.terminal, 1)
    end

    if self.meta.consoleEnable then
        if self.meta.dragging.active then
            self.x = love.mouse.getX() - self.meta.dragging.diffX
            self.y = love.mouse.getY() - self.meta.dragging.diffY
        end
    end
end


function console:textinput(inp)
    if self.meta.consoleEnable then
        if inp ~= self.binds.open then
            self.meta.textCommand = self.meta.textCommand .. inp
        end
    end
end

function console:keypressed(k)
    if k == self.binds.open then
        if self.meta.consoleEnable then
            self.meta.consoleEnable = false
        else
            self.meta.consoleEnable = true
        end
    end
    if self.meta.consoleEnable then
        if k == self.binds.removeChar then
            if #self.meta.textCommand > 0 then
                local byteOffset = utf8.offset(self.meta.textCommand, -1)
                if byteOffset then
                    self.meta.textCommand = string.sub(self.meta.textCommand, 1, byteOffset - 1)
                end
            end
        end
        if k == self.binds.submit then
            console:submit()
        end
        if k == self.binds.previousCommand then
            if #self.meta.commandHistory > 0 then
                self.meta.commandHistorySelection = self.meta.commandHistorySelection - 1
                if self.meta.commandHistorySelection < 1 then
                    self.meta.commandHistorySelection = 1
                end

                if self.meta.commandHistorySelection > #self.meta.commandHistory then
                    self.meta.commandHistorySelection = #self.meta.commandHistory
                end
                self.meta.textCommand = self.meta.commandHistory[self.meta.commandHistorySelection]
            end
        end
        if k == self.binds.nextCommand then
            if #self.meta.commandHistory > 0 then
                self.meta.commandHistorySelection = self.meta.commandHistorySelection + 1
                
                if self.meta.commandHistorySelection < 1 then
                    self.meta.commandHistorySelection = 1
                end

                if self.meta.commandHistorySelection > #self.meta.commandHistory then
                    self.meta.commandHistorySelection = #self.meta.commandHistory
                end
                self.meta.textCommand = self.meta.commandHistory[self.meta.commandHistorySelection]
            end
        end
    end
end

function console:mousepressed(x, y, button)
    if button == 1
    and x > self.x and x < self.x + self.w
    and y > self.y and y < self.y + self.h
    and self.meta.consoleEnable
    then
        self.meta.dragging.active = true
        self.meta.dragging.diffX = x - self.x
        self.meta.dragging.diffY = y - self.y
    else
        self.meta.dragging.active = false
    end
end

function console:mousereleased(x, y, button)
    if button == 1 then
        self.meta.dragging.active = false
    end
end

-------------------------------

--- run a command on the console
---@param cmd string
function console:run(cmd)
    local tkn = tokenize(cmd, " ")
    for _, commands in ipairs(self.meta.commands) do
        if commands.name == tkn[1] then
            table.remove(tkn, 1)
            sucess, error = pcall(commands.func, unpack(tkn))
        end
    end
    tkn = {}
end

function console:submit()
    if self.meta.consoleEnable then
        if self.meta.textCommand ~= "" then
            console:trace(self.meta.textCommand, 0)
            console:run(self.meta.textCommand)
            table.insert(self.meta.commandHistory, self.meta.textCommand)
            self.meta.commandHistoryCurrentItem = #self.meta.commandHistory
            self.meta.textCommand = ""
        end
    end
end

--- Create your custom commands
---@param commandName string
---@param helpDescription string
---@param priorityLevel number
---@param func function
function console:registerCommand(commandName, helpDescription, priorityLevel, func)
    Command = {}
    Command.name = commandName or "defaultCommand"
    Command.priority = priorityLevel or 0
    Command.description = helpDescription or ""
    Command.func = func or function()
        console:trace("please replace it with console:registerCommand()", 2)
    end
    table.insert(self.meta.commands, Command)
end

--- Rebind the console special keys with a table presenting the custom keys
---@param keys table
---@alias keyname
---| "submit"  # ---#DESTAIL 'keys.submit'
---| "open"  # ---#DESTAIL 'keys.open'
---| "removeChar"  # ---#DESTAIL 'keys.removeChar'
---| "previousCommand"  # ---#DESTAIL 'keys.previousCommand'
---| "nextCommand"  # ---#DESTAIL 'keys.nextCommand'
function console:rebind(keys)
    self.binds.submit = keys.submit or "return"
    self.binds.open = keys.open or "f1"
    self.binds.removeChar = keys.removeChar or "backspace"
    self.binds.previousCommand = keys.previousCommand or "up"
    self.binds.nextCommand = keys.nextCommand or "down"
end

--- Check if the console is consoleEnable
---@return boolean
function console:isConsoleEnabled()
    return self.meta.consoleEnable
end

--- Write a message on the console
---@param message string
---@param level number
function console:trace(message, level)
    Message = {}
    Message.type = "trace"
    Message.text = tostring(message) or ""
    if type(level) == "number" then
        if level == nil then
            level = 0
        end
        if level > 3 then
            level = 0
        end
    else
        error("Can't use string as parameter, use a 'number' type", 1)
    end
        if level == 0 then
            Message.levelColor = {255 / 255, 255 / 255, 255 / 255}
        end
        if level == 1 then
            Message.levelColor = {255 / 255, 255 / 255, 0 / 255}
        end
        if level == 2 then
            Message.levelColor = {255 / 255, 0 / 255, 0 / 255}
        end
        if level == 3 then
            Message.levelColor = {0 / 255, 255 / 255, 255 / 255}
        end
    table.insert(self.meta.terminal, Message)
end

--- Special function : with special text formating
---@param message string
function console:announce(message)
    Message = {}
    Message.type = "announce"
    Message.text = tostring(message) or ""
    table.insert(self.meta.terminal, Message)
end

function console:setFont(fontFile, size)
    if size == nil then
        size = 15
    end
    local consoleFont = love.graphics.newFont(fontFile, size)
    self.meta.font = consoleFont
    self.meta.fontSize = size
end

------------- special functions ------------------

function tokenize(inputstr, sep)
    if sep == nil then
        sep = "%s"
    end
    local t = {}
    for str in string.gmatch(inputstr, "([^" .. sep .. "]+)") do
        table.insert(t, str)
    end
    return t
end

-- cpack and cunpack created by mindreframer --
function cpack(t, drop, indent)
	assert(type(t) == "table", "Can only TSerial.pack tables.")
	local s, indent = "{"..(indent and "\n" or ""), indent and math.max(type(indent)=="number" and indent or 0,0)
	for k, v in pairs(t) do
		local tk, tv, skip = type(k), type(v)
		if tk == "boolean" then k = k and "[true]" or "[false]"
		elseif tk == "string" then if string.format("%q",k) ~= '"'..k..'"' then k = '['..string.format("%q",k)..']' end
		elseif tk == "number" then k = "["..k.."]"
		elseif tk == "table" then k = "["..cpack(k, drop, indent and indent+1).."]"
		elseif type(drop) == "function" then k = "["..string.format("%q",drop(k)).."]"
		elseif drop then skip = true
		else error("Attempted to TSerial.pack a table with an invalid key: "..tostring(k))
		end
		if tv == "boolean" then v = v and "true" or "false"
		elseif tv == "string" then v = string.format("%q", v)
		elseif tv == "number" then	-- no change needed
		elseif tv == "table" then v = cpack(v, drop, indent and indent+1)
		elseif type(drop) == "function" then v = "["..string.format("%q",drop(v)).."]"
		elseif drop then skip = true
		else error("Attempted to TSerial.pack a table with an invalid value: "..tostring(v))
		end
		if not skip then s = s..string.rep("\t",indent or 0)..k.."="..v..","..(indent and "\n" or "") end
	end
	return s..string.rep("\t",(indent or 1)-1).."}"
end

function cunpack(s)
	assert(type(s) == "string", "Can only TSerial.unpack strings.")
	assert(loadstring("ctable="..s))()
	local t = ctable
	ctable = nil
	return t
end

function toboolean(str)
    local bool = false
    if string.lower(str) == "true" then
        bool = true
    end
    return bool
end

return console